modules/hal_rpi_pico/CMakeLists.txt

```
@@ -8,60 +10,54 @@ if(CONFIG_HAS_RPI_PICO)
-
-	target_include_directories(boot_stage2 PUBLIC
-		${CMAKE_CURRENT_LIST_DIR}
-		${boot_stage_dir}/asminclude
-		${rp2_common_dir}/pico_platform/include
-		${rp2040_dir}/hardware_regs/include
-		${common_dir}/pico_base/include
-		${ZEPHYR_BASE}/include
-		)
-
-	target_link_options(boot_stage2 PRIVATE
-		"-nostartfiles"
-		"--specs=nosys.specs"
-		"LINKER:--script=${boot_stage_dir}/boot_stage2.ld"
-		)
-
-	# The second stage bootloader is compiled without kconfig definitions.
-	# Therefore, in order to use toolchain.h, it needs to define CONFIG_ARM.
-	target_compile_definitions(boot_stage2 PRIVATE -DCONFIG_ARM=1)
-
-	# Generates a binary file from the compiled bootloader
-	add_custom_target(bootloader_bin ALL DEPENDS boot_stage2.bin)
-	add_custom_command(OUTPUT boot_stage2.bin
-		DEPENDS boot_stage2
-		COMMAND ${CMAKE_OBJCOPY} -Obinary $<TARGET_FILE:boot_stage2> boot_stage2.bin
-		)
+	# The Second Stage Bootloader is only linked to the app that resides
+	# at 0x100. Therefore, only if the app's offset is 0x100, the second
+	# stage bootloader should be compiled.
+	if(CONFIG_RP2_REQUIRES_SECOND_STAGE_BOOT)
+		if(CONFIG_RP2_FLASH_W25Q080)
+			set(flash_type_file boot2_w25q080.S)
+		elseif(CONFIG_RP2_FLASH_GENERIC_03H)
+			set(flash_type_file boot2_generic_03h.S)
+		elseif(CONFIG_RP2_FLASH_IS25LP080)
+			set(flash_type_file boot2_is25lp080.S)
+		elseif(CONFIG_RP2_FLASH_W25X10CL)
+			set(flash_type_file boot2_w25x10cl.S)
+		elseif(CONFIG_RP2_FLASH_AT25SF128A)
+			set(flash_type_file boot2_at25sf128a.S)
+		else()
+			message(FATAL_ERROR "No flash type selected")
+		endif()
 
-	# Checksums the binary, pads it, and generates an assembly file
-	add_custom_target(boot_stage2_asm ALL DEPENDS boot_stage2.S)
-	add_custom_command(OUTPUT boot_stage2.S
-		DEPENDS boot_stage2.bin
-		COMMAND ${Python3_EXECUTABLE} ${boot_stage_dir}/pad_checksum
-		-s 0xffffffff boot_stage2.bin boot_stage2.S)
+		set(rp2_bootloader_prefix ${CMAKE_BINARY_DIR}/bootloader)
+		ExternalProject_Add(
+			second_stage_bootloader
+			PREFIX ${rp2_bootloader_prefix}
+			SOURCE_DIR ${CMAKE_CURRENT_LIST_DIR}/bootloader
+			BINARY_DIR ${rp2_bootloader_prefix}
+		    CONFIGURE_COMMAND
+			    ${CMAKE_COMMAND} -G${CMAKE_GENERATOR}
+			    -S ${CMAKE_CURRENT_LIST_DIR}/bootloader
+			    -B ${rp2_bootloader_prefix}
+		        -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
+			    -DCMAKE_ASM_COMPILER=${CMAKE_ASM_COMPILER}
+		        -DZEPHYR_HAL_RPI_PICO_MODULE_DIR=${ZEPHYR_HAL_RPI_PICO_MODULE_DIR}
+		        -DZEPHYR_BASE=${ZEPHYR_BASE}
+		        -DFLASH_TYPE_FILE=${flash_type_file}
+		        -DRPI_PICO_HAL_DIR=${CMAKE_CURRENT_LIST_DIR}
+		    BUILD_COMMAND
+			    ${CMAKE_COMMAND} --build .
+		    INSTALL_COMMAND "" # No installation needed
+	    )
+		add_dependencies(${ZEPHYR_CURRENT_LIBRARY} second_stage_bootloader)
+		set_source_files_properties(${CMAKE_BINARY_DIR}/bootloader/boot_stage2.S GENERATED)
+		zephyr_library_sources(${CMAKE_BINARY_DIR}/bootloader/boot_stage2.S)
```

when using `BUILD_BYPRODUCTS` this should use same variable description.
```suggestion
		zephyr_library_sources(${rp2_bootloader_prefix}/boot_stage2.S)
```